File: .github/smoke/node/consumer.js
'use strict';

/*
 * Isolated external consumer for the Node.js SDK, mirroring
 * .github/smoke/java/ExternalConsumer.java: uses only the published package's public
 * surface, not a relative path into this repo, network-free, and prints a fixed
 * sentinel the CI workflow greps for. The workflow installs the packed tarball into a
 * fresh throwaway project via npm, then pnpm, then yarn, and runs this same script
 * against each installation — this is what catches pnpm/yarn's stricter
 * phantom-dependency resolution that npm's flat node_modules silently tolerates.
 *
 * The SDK is published as two separate npm packages, not one package with two major
 * version lines — @oppex/integration-sdk (modern, fetch, Node >=18) and
 * @oppex/integration-sdk-legacy (legacy, http/https, Node >=8) — so which one to
 * require is passed in via SDK_PACKAGE_NAME rather than hardcoded, letting this one
 * file cover both instead of needing a second, near-duplicate consumer script.
 *
 * "Network-free" doesn't mean "never actually attempts an HTTP call" — a blank title
 * only ever exercises the validation-failure path, never the transport. Port 1 on
 * loopback reliably refuses the connection immediately (nothing listens there), so a
 * fully valid request pointed at it still reaches the real transport.sendRequest call
 * and fails there, genuinely exercising that path without depending on real network
 * access or touching the actual Oppex service.
 */

var assert = require('assert');
var packageName = process.env.SDK_PACKAGE_NAME || '@oppex/integration-sdk';

process.env.OPPEX_TEST_ENDPOINT_URL = 'http://127.0.0.1:1';

var sdk = require(packageName);

function main() {
  var IncidentClient = sdk.IncidentClient;
  var Severity = sdk.Severity;

  assert.strictEqual(typeof IncidentClient, 'function', 'IncidentClient export missing');
  assert.strictEqual(Severity.MEDIUM, 3, 'unexpected severity mapping');

  // Deliberately wrong-looking credentials — the point of this first check is
  // reaching the real transport call and having THAT fail (connection refused), not a
  // credentials-shaped validation error.
  var badClient = new IncidentClient({ apiKey: 'wrong-api-key', serviceKey: 'wrong-service-key' });

  return badClient
    .sendIncident({ title: 'valid title', source: 'github-actions', severity: Severity.MEDIUM })
    .then(function (response) {
      assert.strictEqual(response.successful, false, 'a real transport failure must resolve failed, not throw');
      assert.strictEqual(response.code, -1, 'a connection failure must resolve with code -1');
      return badClient.close();
    })
    .then(function () {
      var client = new IncidentClient({
        apiKey: 'external-consumer-api-key',
        serviceKey: 'external-consumer-service-key',
      });

      // Invalid request (blank title) resolves via validation before any HTTP
      // attempt, exercising the "never throws" contract from that side too.
      return client
        .sendIncident({ title: '', source: 'github-actions', severity: Severity.MEDIUM })
        .then(function (response) {
          assert.strictEqual(response.successful, false, 'invalid request must resolve failed, not throw');
          return client.close();
        });
    })
    .then(function () {
      console.log('EXTERNAL_CONSUMER_OK node=' + process.version);
    });
}

main().catch(function (err) {
  console.error(err);
  process.exit(1);
});

File: .github/smoke/go/consumer.go
// Command consumer exercises the published Go SDK surface from outside its own
// module, mirroring .github/smoke/java/ExternalConsumer.java: only supported
// API, network-free, and a fixed sentinel the workflow greps for.
//
// "Network-free" does not mean "never attempts an HTTP call". Port 1 on loopback
// refuses the connection immediately, so a fully valid request pointed at it
// still reaches the real transport and fails there, genuinely exercising that
// path without touching the actual Oppex service.
package main

import (
	"context"
	"errors"
	"fmt"
	"os"
	"runtime"
	"strings"

	"github.com/Oppex-AI/oppex-integration/golang/oppex"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	fmt.Printf("EXTERNAL_CONSUMER_OK go=%s\n", runtime.Version())
}

func run() error {
	if oppex.DefaultEndpoint != "https://api.oppex.ai/api/v1/incident/post" {
		return fmt.Errorf("unexpected endpoint %q", oppex.DefaultEndpoint)
	}
	if oppex.SeverityMedium != 3 {
		return fmt.Errorf("unexpected severity mapping %d", oppex.SeverityMedium)
	}

	if err := checkRealTransportFailure(); err != nil {
		return err
	}
	if err := checkValidation(); err != nil {
		return err
	}
	return checkServiceRouting()
}

// checkRealTransportFailure reaches the actual transport and requires it to fail
// with a connection refusal, not a validation error.
func checkRealTransportFailure() error {
	os.Setenv("OPPEX_TEST_ENDPOINT_URL", "http://127.0.0.1:1")
	defer os.Unsetenv("OPPEX_TEST_ENDPOINT_URL")

	client, err := oppex.New(oppex.Config{APIKey: "wrong-api-key", ServiceKey: "wrong-service-key"})
	if err != nil {
		return fmt.Errorf("creating the client: %w", err)
	}
	defer client.Close()

	_, err = client.Post(context.Background(), oppex.IncidentRequest{
		Title: "valid title", Source: "github-actions", Severity: oppex.SeverityMedium,
	})

	var failure *oppex.IncidentError
	if !errors.As(err, &failure) {
		return fmt.Errorf("a refused connection must be an *oppex.IncidentError, got %v", err)
	}
	if failure.StatusCode != oppex.NoStatusCode {
		return fmt.Errorf("a refused connection must carry NoStatusCode, got %d", failure.StatusCode)
	}
	// Five retries plus the first attempt, so the message carries the count.
	if !strings.Contains(failure.Error(), "attempts") {
		return fmt.Errorf("an exhausted network failure must report its attempts: %v", failure)
	}
	return nil
}

func checkValidation() error {
	client, err := oppex.New(oppex.Config{APIKey: "external-consumer-api-key", ServiceKey: "k"})
	if err != nil {
		return fmt.Errorf("creating the client: %w", err)
	}
	defer client.Close()

	// A blank title fails before any HTTP attempt.
	_, err = client.Post(context.Background(), oppex.IncidentRequest{
		Title: "", Source: "github-actions", Severity: oppex.SeverityMedium,
	})
	if !errors.Is(err, oppex.ErrInvalidRequest) {
		return fmt.Errorf("a blank title must be rejected, got %v", err)
	}

	if err := client.Close(); err != nil {
		return fmt.Errorf("closing the client: %w", err)
	}
	if _, err := client.Post(context.Background(), oppex.IncidentRequest{
		Title: "t", Source: "github-actions", Severity: oppex.SeverityLow,
	}); !errors.Is(err, oppex.ErrClientClosed) {
		return fmt.Errorf("a post after Close must report a closed client, got %v", err)
	}
	return nil
}

// checkServiceRouting exercises the precondition, which fails before any network
// call.
func checkServiceRouting() error {
	client, err := oppex.New(oppex.Config{APIKey: "external-consumer-api-key"})
	if err != nil {
		return fmt.Errorf("creating the client: %w", err)
	}
	defer client.Close()

	_, err = client.PostWithServiceRouting(context.Background(), oppex.IncidentRequest{
		Title: "Service routing test", Source: "github-actions",
		Severity: oppex.SeverityLow, ServiceKey: "external-consumer-service-key",
	})
	if !errors.Is(err, oppex.ErrInvalidRequest) {
		return fmt.Errorf("service routing must refuse a request service key, got %v", err)
	}
	return nil
}
